core/src/main/java/com/linecorp/armeria/common/logging/DefaultRequestLog.java

```diff
@@ -918,19 +959,40 @@ private void endRequest0(@Nullable Throwable requestCause, long requestEndTimeNa
         }
 
         if (name == null) {
+            String newServiceName = null;
             String newName = null;
-            final RpcRequest rpcReq = ctx.rpcRequest();
+            RpcRequest rpcReq = ctx.rpcRequest();
+            if (rpcReq == null && requestContent instanceof RpcRequest) {
+               rpcReq = (RpcRequest) requestContent;
+            }
+
             if (rpcReq != null) {
+                newServiceName = rpcReq.serviceType().getName();
                 newName = rpcReq.method();
-            } else if (requestContent instanceof RpcRequest) {
-                newName = ((RpcRequest) requestContent).method();
-            }
+            } else if (ctx instanceof ServiceRequestContext) {
+                final ServiceConfig config = ((ServiceRequestContext) ctx).config();
+                newServiceName = config.defaultServiceName();
+                if (newServiceName == null) {
+                    newServiceName = config.route().meterTag();
+                }
 
-            if (newName == null && ctx instanceof ServiceRequestContext) {
-                newName = ((ServiceRequestContext) ctx).config().defaultLogName();
+                newName = config.defaultLogName();
+                if (newName == null) {
+                    newName = ctx.method().name();
+                }
```

Comment:
Should we set the fallback serviceName and name for the client side as well? Then we could make the two properties non-null.

Reply:
> Should we set the fallback serviceName and name for the client side as well? 

Yes. That was one of the considering points. Let me apply the names to client-side.

Reply:
On second thought, there is inconsistency in `serviceName`:
- The server-side requests use a path pattern for fallback `serviceName`. (eg. `/users/:id`)
- The client-side could not get a path pattern except for Retrofit. Is it good to use path? (eg. `/users/123`)

Reply:
How about this:

- Add `Service.unwrap()` that always unwraps at least once, so that we can get the innermost service instance,
- and then use its class name as the fallback service name
- For the client-side, how about falling back to authority? 🤔

Reply:
On second thought, I guess it doesn't make much sense to provide the default service name on the client side except for Retrofit.

Reply:
> The client-side could not get a path pattern except for Retrofit. Is it good to use a path as a fallback serviceName? (eg. /users/123) 🧐

I don't think so. It will increase the number of tags too much. Sounds like `serviceName()` should stay nullable, while `name()` could become non-null. WDYT?

Reply:
>  Sounds like serviceName() should stay nullable, while name() could become non-null. WDYT?

Agreed. It sounds good that `serviceName()` is nullable in clien-side.

> Add Service.unwrap() that always unwraps at least once, so that we can get the innermost service instance.

That's a good idea. Let me try it.